design-documents/sessions-only-for-html.md

```diff
@@ -0,0 +1,20 @@
+### Magento\Framework\Session\SessionManagerInterface should be used only in controllers and blocks
+#### Why?
+Web APIs (REST (by definition), SOAP, GraphQL) do not employ sessions,
+relying on sessions in classes that are not directly linked to preparing HTML
+may cause sessions being used during Web API requests unintentionally causing
+unexpected behaviour during and negative performance impact.
+ 
+#### Current situation
+Sessions are being used in model classes.
+ 
+#### What has to be done
+Add PHPMD rule which will be triggered when a class has a constructor parameter of type
+SessionManagerInterface (and any of the implementations) and is not an instance of
+ActionInterface or AbstractBlock.
```

Comment:
1. Also a plugin for a block/controller should be OK to use sessions.
1. Would be good to expand the rule to cover "6.2.2. Request, Response, Session, Store Manager and Cookie objects MUST be used only in the Presentation layer.". Can be implemented as separate rules.
1. Also, data providers for UI components are part of presentation layer, so they can depend on session.

Reply:
1. makes sense, but I don't know how a block plugin can be identified in a PHPMD rule. Controller plugin can be sort of identified by having before/after/aroundExecute/Dispatch but I'm not sure if that would be correct. I think it would be OK if devs would just have to put SuppressWarnings for those cases
2. Ok, I'll update the proposal
3. Sure, instances of Magento\Framework\View\Element\UiComponent\DataProvider\Document and Magento\Framework\View\Element\UiComponent\DataProvider\DataProviderInterface will be ignored by the rule then

Reply:
1. The plugin has `$subject` which can go through the same checks on whether it's a block or controller.